pygpcca/_gpcca.py

```diff
@@ -1001,14 +1013,19 @@ def optimize(
             chi_list.append(chi)
             rot_matrix_list.append(rot_matrix)
 
-        opt_idx = np.argmax(crispness_list)
+        if np.any(np.array(crispness_list) > 0.0):
+            opt_idx = np.argmax(crispness_list)
+        else:
+            raise ValueError("Clustering wasn't successful. Try different cluster numbers.")
         self._m_opt = min(m_list) + opt_idx
         self._chi = chi_list[opt_idx]
         self._rot_matrix = rot_matrix_list[opt_idx]
-        self._crispness = crispness_list[opt_idx]
+        self._crispness = np.array(crispness_list)
+        self._crispness_opt = crispness_list[opt_idx]
         self._X = self._p_X[:, : self._m_opt]
         self._R = self._p_R[: self._m_opt, : self._m_opt]
-        self._eigenvalues = self._p_eigenvalues[: self._m_opt]
+        self._top_eigenvalues = self._p_eigenvalues[: self._m_opt]
+        self._eigenvalues = self._p_eigenvalues[min(m_list) - 1 : max(m_list)]
```

Comment:
Why wouldn't this just be all computed eigenvalues? Why do we not start from zero here?

Reply:
Just for aesthetic reasons - since, if we are optimizing for cluster numbers `m in [m_min, m_max]` I would expect to get the eigenvalues matching those numbers.
But for practical reasons it might be more useful to get `self._eigenvalues = self._p_eigenvalues[ : max(m_list)]`.

Reply:
Done [here](https://github.com/msmdev/pyGPCCA/pull/13/commits/5f9bd45dcf3ea37c64987aba5a22a53b30f199e8).